lib/db_connection.ex

```diff
@@ -785,65 +894,382 @@ defmodule DBConnection do
     * `:queue` - Whether to block waiting in an internal queue for the
     connection's state (boolean, default: `true`)
     * `:timeout` - The maximum time that the caller is allowed the
-    to hold the connection's state (default: `15_000`)
-    * `:log` - A function to log information about begin, commit and rollback
-    calls made as part of the transaction, either a 1-arity fun,
-    `{module, function, args}` with `DBConnection.LogEntry.t` prepended to
-    `args` or `nil`. See `DBConnection.LogEntry` (default: `nil`)
+    to hold the connection's state (ignored when using a locked connection,
+    default: `15_000`)
+    * `:log` - A function to log information about a call, either
+    a 1-arity fun, `{module, function, args}` with `DBConnection.LogEntry.t`
+    prepended to `args` or `nil`. See `DBConnection.LogEntry` (default: `nil`)
 
   The pool and connection module may support other options. All options
-  are passed to `handle_begin/2`, `handle_commit/2` and
-  `handle_rollback/2`.
+  are passed to `handle_begin/2`.
+
+  See `commit/2` and `rollback/2`.
 
   ### Example
 
-      {:ok, res} = DBConnection.transaction(conn, fn(conn) ->
-        DBConnection.execute!(conn, "SELECT id FROM table", [])
-      end)
+      {:ok, result} = DBConnection.begin(conn)
+      try do
+        DBConnection.execute(conn, "SELECT * FROM table", [])
+      after
+        DBConnection.commit(conn)
+      end
   """
-  @spec transaction(conn, (conn -> result), opts :: Keyword.t) ::
-    {:ok, result} | {:error, reason :: any} when result: var
-  def transaction(conn, fun, opts) do
-    case run(conn, &run_transaction/5, fun, meter(opts), opts) do
-      {:commit, {:ok, res, return, meter}} ->
-        log(meter, :transaction, :commit, nil, {:ok, res})
-        {:ok, return}
-      {:rollback, {:ok, res, reason, meter}} ->
-        log(meter, :transaction, :rollback, nil, {:ok, res})
-        {:error, reason}
-      {:rollback, {:ok, res, kind, reason, stack, meter}} ->
-        log(meter, :transaction, :rollback, nil, {:ok, res})
-        :erlang.raise(kind, reason, stack)
-      {query, {:error, err, meter}} when query != nil ->
-        log(meter, :transaction, query, nil, {:error, err})
+  @spec begin(conn, opts :: Keyword.t) :: {:ok, result} | {:error, Exception.t}
+  def begin(conn, opts \\ []) do
+    conn
+    |> run(&run_begin/4, meter(opts), opts)
+    |> log(:begin, :begin, nil)
```

Comment:
As this function would never be used with a pool (unless crazy) we could merge it with `checkout_begin`, such that if the pool is used it performs the current `checkout_begin`. The `pool` version implemented here could be emulated by immediately checking in. This would mean `begin(conn, opts) :: {:ok, conn, result}`.